### PR TITLE
deprecate config-magic in favor of json configuration stuff

### PR DESCRIPTION
**Pull request by clintropolis**
### Description
This PR aims to deprecate usage of config-magic (`ConfigProvider`) in favor of using `JsonConfigProvider` everywhere. In the process I have made the config paths a bit more sane (one of the reasons to replace `ConfigProvider` is because of how easily devs can make paths that don't make much structural sense due to the free-form nature of how config paths are defined).

The main users are `DruidProcessingConfig` and `DruidCoordinatorConfig`, the former of which has been reworked to be simple jackson annotated objects which are injected with a fallback to support legacy config-magic paths as needed. After these changes are released, we can consider finally deleting config-magic in a future release.

For `DruidProcessingConfig`, the structure was mostly ok, but I have flattened the broker parallel merge config into just living at `druid.processing.merge` (i guess we could change this, i don't feel very strongly), and split it into a totally separate config from the rest of `DruidProcessingConfig` since it is only used by brokers for `CachingClusteredClient`.

For `DruidCoordinatorConfig` WIP (i haven't done anything to it yet, but it definitely needs reworked...)

`MonitorSchedulerConfig` was also using config-magic annotations, but was no longer using `ConfigProvider` and so seemed unnecessary so I have just flattened it into `DruidMonitorSchedulerConfig` which has moved to druid-processing, because it was already wired up to `JsonConfigProvider`.

#### Release note
Broker parallel merge config options at paths `druid.processing.merge.pool.*` and `druid.processing.merge.task.*` have been flattened to use the `druid.processing.merge` prefix. Note that the legacy paths will still function, but will issue a warning log noting that these configs have been deprecated and will be removed in a future release, along with the new paths which should be used to continue configuring these values.

<hr>

This PR has:

- [ ] been self-reviewed.
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Review by kfaraz — COMMENTED** [review]
@clintropolis , the changes look good to me overall. I have left some minor comments.

As for `DruidCoordinatorConfig`, I am looking into how we can clean up the structure. I will share my thoughts here soon.

But for now, I think we should merge the current PR (as it has already touched several files) and tackle `DruidCoordinatorConfig` in a separate PR of its own.

**Comment by kfaraz** [comment]
For `DruidCoordinatorConfig`, I think the following structure should work:

Existing class `DruidCoordinatorConfig` bound to `druid.coordinator`
- `startDelay`
- `period`
- `kill`: new class `CoordinatorKillConfigs`
  - `unused`: new class `KillUnusedSegmentsConfig`
    - `period`
    - `durationToRetain`
    - `ignoreDurationToRetain`
    - `maxSegments`
  - `supervisor`: new class `MetadataCleanupConfig` for this and other cleanups below
    - `period`
    - `durationToRetain`
  - `audit`
    - `period`
    - `durationToRetain`
  - `datasource`
    - `period`
    - `durationToRetain`
  - `rule`
    - `period`
    - `durationToRetain`
  - `compaction`
    - `period`
- `loadqueuepeon`: new class `LoadQueuePeonConfig`
  - `type`
  - `loadTimeout` (replacing `druid.coordinator.load.timeout`)
  - `http`: new class `HttpLoadQueuePeonConfig`
    - `batchSize`
    - `repeatDelay`
    - `hostTimeout`

Deprecated paths:
new class `LegacyKillUnusedSegmentsConfig` which has the following paths
- `druid.coordinator.kill.period`
- `druid.coordinator.kill.durationToRetain`
- `druid.coordinator.kill.ignoreDurationToRetain`
- `druid.coordinator.kill.maxSegments`

Removed paths:
- `druid.coordinator.curator.loadqueuepeon.numCallbackThreads`
Curator loading is already deprecated and will be removed before the next release. We can remove this config and hardcode the only usage in `CliCoordinator` to 2 for the time being.
- `druid.coordinator.load.timeout`
This can be replaced by `druid.coordinator.loadqueuepeon.loadTimeout`. The older config can just be dropped as I have never seen any cluster configure this value to anything other than the default (15 minutes). Also, I have other changes planned for this area which are going to further refine this config.

---

@clintropolis , let me know what you think.

File: processing/src/main/java/org/apache/druid/query/DruidProcessingConfig.java [hunk 1]
```diff
@@ -19,45 +19,85 @@
...
+      @JsonProperty("indexes") DruidProcessingIndexesConfig indexes
+  )
+  {
+    this.formatString = formatString == null ? "processing-%s" : formatString;
+    this.numThreads = numThreads == null
+                      ? Math.max(JvmUtils.getRuntimeInfo().getAvailableProcessors() - 1, 1)
+                      : numThreads;
+    this.numMergeBuffers = numMergeBuffers == null ? Math.max(2, this.numThreads / 4) : numMergeBuffers;
+    this.fifo = fifo == null || fifo;
+    this.tmpDir = tmpDir == null ? System.getProperty("java.io.tmpdir") : tmpDir;
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/query/DruidProcessingConfig.java]: cool, didn't know that existed, changed to use this where appropriate

File: processing/src/main/java/org/apache/druid/java/util/metrics/DruidMonitorSchedulerConfig.java [hunk 2]
```diff
@@ -17,24 +17,23 @@
...
+public class DruidMonitorSchedulerConfig
 {
   @JsonProperty
   private String schedulerClassName = BasicMonitorScheduler.class.getName();
 
   @JsonProperty
   private Period emissionPeriod = new Period("PT1M");
 
+  @JsonProperty
   public String getSchedulerClassName()
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/java/util/metrics/DruidMonitorSchedulerConfig.java]: they don't currently get serialized so its not really necessary, not sure why some of the getters here had them, removed them all to make consistent

File: processing/src/main/java/org/apache/druid/java/util/metrics/DruidMonitorSchedulerConfig.java [hunk 3]
```diff
@@ -46,7 +45,6 @@ public Period getEmissionPeriod()
     return emissionPeriod;
   }
 
-  @Override
   public Duration getEmitterPeriod()
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/java/util/metrics/DruidMonitorSchedulerConfig.java]: renamed, though the other method wasn't actually used which became apparent after removing the json annotation, so i removed it too

File: processing/src/main/java/org/apache/druid/query/DruidProcessingConfig.java [hunk 4]
```diff
@@ -19,45 +19,85 @@
...
-  public static final HumanReadableBytes DEFAULT_PROCESSING_BUFFER_SIZE_BYTES = HumanReadableBytes.valueOf(-1);
-  public static final int MAX_DEFAULT_PROCESSING_BUFFER_SIZE_BYTES = 1024 * 1024 * 1024;
-  public static final int DEFAULT_MERGE_POOL_AWAIT_SHUTDOWN_MILLIS = 60_000;
-  public static final int DEFAULT_INITIAL_BUFFERS_FOR_INTERMEDIATE_POOL = 0;
-
-  private AtomicReference<Integer> computedBufferSizeBytes = new AtomicReference<>();
-
-  @Config({"druid.computation.buffer.size", "${base_path}.buffer.sizeBytes"})
-  public HumanReadableBytes intermediateComputeSizeBytesConfigured()
+  public static long computeMaxMemoryFromMaxHeapSize()
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/query/DruidProcessingConfig.java]: inlined

File: server/src/main/java/org/apache/druid/guice/RouterProcessingModule.java [hunk 5]
```diff
@@ -53,16 +53,17 @@ public class RouterProcessingModule implements Module
...
   }
 
   @Provides
   @ManageLifecycle
   public QueryProcessingPool getProcessingExecutorPool(DruidProcessingConfig config)
   {
-    if (config.getNumThreadsConfigured() != ExecutorServiceConfig.DEFAULT_NUM_THREADS) {
-      log.error("numThreads[%d] configured, that is ignored on Router", config.getNumThreadsConfigured());
+    if (config.isNumThreadsConfigured()) {
+      log.error("numThreads[%d] configured, that is ignored on Router", config.getNumThreads());
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply server/src/main/java/org/apache/druid/guice/RouterProcessingModule.java]: changed to warn, iirc the idea was that these configs have no effect on router so logging about it makes it obvious that a config change didn't actually change anything.  I don't really feel strongly either way, though i do see some merit to getting feedback when a config is specified that has no effect, however we aren't doing it very consistently

File: processing/src/main/java/org/apache/druid/query/DruidProcessingBufferConfig.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.druid.java.util.common.HumanReadableBytes;
+
+import javax.annotation.Nullable;
+
+public class DruidProcessingBufferConfig
+{
+  public static final HumanReadableBytes DEFAULT_PROCESSING_BUFFER_SIZE_BYTES = HumanReadableBytes.valueOf(-1);
+  public static final int MAX_DEFAULT_PROCESSING_BUFFER_SIZE_BYTES = 1024 * 1024 * 1024;
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/query/DruidProcessingBufferConfig.java:33]: hmm, it is only a default though, if the user specifies a buffer size we use that instead, so this just specifies the max size we use if the user didn't define the buffer size explicitly and we are on a JVM that supports auto computing the size

[the remark this answers is not in the document]
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/query/DruidProcessingBufferConfig.java:33]: iow, i didn't change this for now, but am happy to do so if you feel more strongly about it

**Comment by clintropolis** [comment]
>For DruidCoordinatorConfig, I think the following structure should work:
> ...

That seems pretty reasonable to me, since we are switching to JSON config, and a ton of those configs have a `period` and 
`durationToRetain`, it feels like there is some common class that should be able to be re-used for most of those configs.

>druid.coordinator.curator.loadqueuepeon.numCallbackThreads
Curator loading is already deprecated and will be removed before the next release. We can remove this config and hardcode the only usage in CliCoordinator to 2 for the time being.

If for some reason we don't get this removed prior to the next release, we should probably fix it to be consistent with the http load queue peon config, and make its path be `druid.coordinator.loadqueuepeon.curator.numCallbackThreads`

File: processing/src/main/java/org/apache/druid/query/DruidProcessingBufferConfig.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (kfaraz) [reply processing/src/main/java/org/apache/druid/query/DruidProcessingBufferConfig.java:32]: No, it is fine. We can leave as it is.

File: server/src/main/java/org/apache/druid/guice/LegacyBrokerParallelMergeConfigModule.java [hunk 7]
```diff
@@ -19,17 +19,17 @@
...
 
-import com.google.common.collect.ImmutableMap;
 import com.google.inject.Binder;
 import com.google.inject.Module;
-import org.apache.druid.query.DruidProcessingConfig;
+import org.apache.druid.query.LegacyBrokerParallelMergeConfig;
 
-public class DruidProcessingConfigModule implements Module
+@Deprecated
+public class LegacyBrokerParallelMergeConfigModule implements Module
```
[the remark this answers is not in the document]
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/guice/LegacyBrokerParallelMergeConfigModule.java:34]: Thanks!

**Review by kfaraz — APPROVED** [review]

**Pull request merged**